### ISSUE TITLE
Socket/Process events break after upgrade from 2.11.0 -> 3.2.6

### ISSUE THREAD

**Issue by clong** [facebook/osquery#4615]
osquery version: 3.2.6

OS: CentOS6 and Centos7
Background: We had many CO6 and CO7 hosts running auditing and after an upgrade to 3.2.6, they continued logging events from standard differential queries but all socket and process events no longer generated logs.

Debugging steps:
After hours of debugging, the only consistent fix was renaming osquery.db and restarting the daemon. Audit is configured and running correctly on all the hosts, and no combination of flag changing or anything else helped solve the problem until the database was recreated. Auditing also works fine in osqueryi since it uses an ephemeral database.

Oddly enough, I still received warning events for socket_events overflowing, which makes me think that records are being stored, but cannot be queried.

For this reason, I believe something must have changed with how socket/process events are either being stored or queried from RocksDB, leading them to not be logged in the results.

**Comment by clong** [comment]
While trying to repro this in a VM, I'm only able to get to a similar state, but not quite identical:

VM Repro (CentOS 7 VM):
Install osquery 2.6 with process and socket auditing enabled
Install osquery 2.11.0 over the 2.6 installation, keeping auditing enabled
Install osquery 3.2.6 over 2.11.0 and auditing will cease to work. 

However in the VM, I was able to get auditing working again by simply restarting the daemon once more.

In the affected systems, I have tried:
- Restarting osquery multiple times
- Changing audit flag values
- Changing syslog/filesystem logging plugins
- Flipping auditing on and off while osquery is running (`auditctl -e 0; auditctl -e 1`)

The only thing that seems to work is renaming the db and bouncing the daemon

**Comment by jaredledvina** [comment]
@clong - What's the output of `auditctl -l` and `auditctl -s` w/ v3.2.6?

**Comment by clong** [comment]
Bad news - downgrading is not a reliable fix here. It worked on a handful of sample hosts that I tested it on, but the only way we were able to recover our fleet was to downgrade to 2.11.2 and then delete `osquery.db`. I would really like to understand the root cause of why this happened if anyone has some spare cycles to troubleshoot with me -- I'm really unfamiliar with RocksDB and I'm fairly certain something must have changed with the way records are stored/queried at some point

**Comment by theopolis** [comment]
I'm guessing this is related to DB migration logic.

**Comment by fmanco** [comment]
I would bet on the same thing, I think we don't have migration logic for the event records.

@clong are you able to repro this consistently? Can you get a dump of the DB before and after migration? Thanks.

**Comment by clong** [comment]
@fmanco yeah, I can reproduce. What osquery version did the DB migration logic change? I'm not sure which version is considered "after".

**Comment by fmanco** [comment]
The versions you reported on this issue would be good, i.e. `2.11.0` for before and `3.2.6` (or higher) for after.

**Comment by npamnani** [comment]
@clong , the issue I observe is that if there are events stored in the backend events table(staging tables) those events are not accessible after the upgrade from 2.11.0 to 3.2.6. Because of the following change -
 class AuditEventPublisher final
     : public EventPublisher<AuditSubscriptionContext, AuditEventContext> {
-  DECLARE_PUBLISHER("audit");
+  DECLARE_PUBLISHER("auditeventpublisher");
New generated events I can query.

**Comment by clong** [comment]
@fmanco I just sent you the DB's from my VM via Slack

**Comment by fmanco** [comment]
Thanks @clong .

**Comment by fmanco** [comment]
Hi @clong sorry for the delay. I was just looking at the DBs you sent me. We need to add migration to rename old keys from `data.audit.*` to `data.auditeventpublisher.*`. I assume you can query new events correct?

**Comment by clong** [comment]
@fmanco can you elaborate by what you mean by:
> I assume you can query new events correct?

When our hosts get in this state, they simply no longer log any process or socket events until we reset the DB. I'm not sure what you mean by "new" exactly.

**Comment by fmanco** [comment]
@clong I thought the problem was similar to @npamnani 's, i.e. you wouldn't be able to query old events but new ones you could. On the DBs you sent me, post migration I see events with the new key, so new events are being stored and the problem is somehow on the querying part. I need to debug more as I'm not being able to repro.

**Comment by clong** [comment]
@fmanco if you want to hop on a video conference call next week I can walk you through the repro of the problem and maybe we can dig into the problem in a bit more detail? Feel free to ping me on slack and let me know

**Comment by clong** [comment]
Using osqueryi and pointing the `--database_path` flag to the DB shows new events are being added, but they are not being logged by osqueryd.

**Comment by npamnani** [comment]
@clong , do you see the following message in your logs where it is failing -
"Opening RocksDB failed: Continuing with read-only support"

**Comment by npamnani** [comment]
If you find this error, then events are disabled no matter if you have mentioned all the required command line options.

**Comment by fmanco** [comment]
I was finally able to repo, thanks @clong for all the help.

**Problem**:

With the move to the new audit implementation all DB keys but `optmize.*` and `optmize_eid.*` change from `*.audit.*` to `*.auditeventpublisher.*`. I didn't debug this in too much detail for the sake of time but this somehow screws up the way the table decides which events to show when you select from it. New events are still being added to the DB under the new key names.

If you wait long enough it seems the problem solves itself eventually (at least for some instances) but I couldn't reliably determine why or under which conditions.

**Solution**
We need to add the key migration to our migration code. For the time being I wrote a small bash script which executes this migration:

```
#!/bin/bash

DB="/var/osquery/osquery.db"
CF="queries"

LDB="ldb --db=${DB} --column_family=${CF}"


for key in $(${LDB} scan --no_value 2>/dev/null); do
    nkey=$(echo ${key} | sed -s 's/\.audit\./\.auditeventpublisher\./')
    value=$(${LDB} get ${key})

    ${LDB} delete ${key}
    ${LDB} put "${nkey}" "${value}"
done
```

You'll need to build ldb out of the rocksdb version we use for osquery. I've tried different versions which corrupt the DB when writing to it.

Please give this a try and let me know whether it fixes it for you.

**Comment by clong** [comment]
@fmanco this has been running over an hour on a single DB. The DB I'm running it against is 66M.  I'll let you know if things start working again after it finishes running.

**Comment by fmanco** [comment]
That's a lot of events 😮 . Yeah, this isn't really an optimized way to do this, C app would be better. Let me know once it's done and I'll put this to the migration code on core.

**Comment by clong** [comment]
Hey @fmanco, I just finished testing this on 6 different hosts that were experiencing this issue and that script fixed it on all of them. Thanks for all the help here - hopefully we can fit that migration logic into osquery for a 3.3.0 release!
